torch_np/_detail/implementations.py

```diff
@@ -95,6 +225,129 @@ def diff(a_tensor, n=1, axis=-1, prepend_tensor=None, append_tensor=None):
     return result
 
 
+# #### concatenate and relatives
+
+
+def concatenate(tensors, axis=0, out=None, dtype=None, casting="same_kind"):
+    # np.concatenate ravels if axis=None
+    tensors, axis = _util.axis_none_ravel(*tensors, axis=axis)
+
+    if out is not None or dtype is not None:
+        # figure out the type of the inputs and outputs
+        out_dtype = out.dtype.torch_dtype if dtype is None else dtype
+
+        # cast input arrays if necessary; do not broadcast them agains `out`
+        tensors = _util.cast_dont_broadcast(tensors, out_dtype, casting)
+
+    try:
+        result = torch.cat(tensors, axis)
+    except (IndexError, RuntimeError):
+        raise _util.AxisError
```

Comment:
What if there's a shape error?

Reply:
Not sure what you mean. Numpy raises np.AxisError, with
```
In [17]: np.AxisError.__mro__
Out[17]: 
(numpy.AxisError,
 ValueError,
 IndexError,
 LookupError,
 Exception,
 BaseException,
 object)
```

here we raise an AxisError which is

```
class AxisError(ValueError, IndexError):
    pass
```

Reply:
My point is that `np.concatenate` does throw other errors as well:
```python
>>> import numpy as np
>>> x = np.ndarray([1,2,3])
>>> y = np.ndarray([1,2,3,5])
>>> np.concatenate([x,y], axis=0)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<__array_function__ internals>", line 200, in concatenate
ValueError: all the input arrays must have same number of dimensions, but the array at index 0 has 3 dimension(s) and the array at index 1 has 4 dimension(s)
```
I think it's fine not to be too fine with errors just yet, as we may sometimes mask other errors, ad that's not nice.

Reply:
Yes. For the time being though, I think we're OK with just raising the right error class.

Reply:
My point is that sometimes it's not even the right error class. In a similar case
```python
>>> torch.cat([torch.arange(4).unsqueeze(1), torch.arange(5).unsqueeze(1)], dim=1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: Sizes of tensors must match except in dimension 1. Expected size 4 but got size 5 for tensor number 1 in the list.
```
you'd be swallowing that error and returning an incorrect `AxisError` with no message, which is super confusing.

I think the treatment of errors is best left for the future. The main reasons being:
- Simplicity of our code
- It's super difficult to be consistent here
- One of our starting goals was not to fully imitate NumPy, but to imitate **correct** NumPy programs, so that leaves errors pretty much out of scope.

Reply:
Here specifically, numpy raises a ValueError, 

```
[31]: np.concatenate((np.arange(4)[:, None], np.arange(3)[:, None]), axis=1)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Input In [31], in <cell line: 1>()
----> 1 np.concatenate((np.arange(4)[:, None], np.arange(3)[:, None]), axis=1)

File <__array_function__ internals>:200, in concatenate(*args, **kwargs)

ValueError: all the input array dimensions except for the concatenation axis must match exactly, but along dimension 0, the array at index 0 has size 4 and the array at index 1 has size 3
```

so a try-except catches it. 

That said, we're on the same page ISTM, on leaving fine details of having nice error messages to a yet unspecificied later point in the lifecycle of this project.